### PR TITLE
Return Set<PublishDiagnosticsParams> in WorkspaceCompiler#compile

### LINKED ISSUE #101: WorkspaceCompiler#compile should return Set<PublishDiagnosticsParams> instead of Set<Diagnostic>
This would fix the ugly exception being caught when the source of the diagnostic is default (groovyc), and would stop relying on source being an URI, which it doesn't have to be.

https://github.com/palantir/groovy-language-server/blob/develop/groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerState.java#L115

### PR DESCRIPTION
**Pull request by natgabb**
Closes #101

### CODE REVIEW

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java [hunk 1]
```diff
@@ -230,15 +233,18 @@ private void resetCompilationUnit() {
...
-    private Set<Diagnostic> parseErrors(ErrorCollector collector) {
-        Set<Diagnostic> diagnostics = Sets.newHashSet();
+    private Set<PublishDiagnosticsParams> parseErrors(ErrorCollector collector) {
+        Map<URI, PublishDiagnosticsParamsBuilder> diagnosticsByFile = Maps.newHashMap();
+
         for (int i = 0; i < collector.getWarningCount(); i++) {
             WarningMessage message = collector.getWarning(i);
-
-            diagnostics.add(new DefaultDiagnosticBuilder(message.getMessage(), DiagnosticSeverity.Warning).build());
+            diagnosticsByFile.computeIfAbsent(workspaceRoot.toUri(),
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java:241]: so sad context is protected T^T

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java [hunk 2]
```diff
@@ -256,11 +262,13 @@ private void resetCompilationUnit() {
...
                 message.write(writer);
+                uri = workspaceRoot.toUri();
                 diagnostic = new DefaultDiagnosticBuilder(data.toString(), DiagnosticSeverity.Error).build();
             }
-            diagnostics.add(diagnostic);
+            diagnosticsByFile.computeIfAbsent(uri, (value) -> new PublishDiagnosticsParamsBuilder().uri(uri.toString()))
+                    .diagnostic(diagnostic);
         }
-        return diagnostics;
+        return diagnosticsByFile.values().stream().map(builder -> builder.build()).collect(Collectors.toSet());
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java]: .map(PublishDiagnosticsParamsBuilder::build)?

**Comment by jiahuijiang** [comment]
+1 after comments

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java (hunk 1 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java:241]: :(

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java (hunk 2 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyWorkspaceCompiler.java]: fixed

**Pull request merged**